From c4fdd252cabd0d9aa9792209be6221f07e6bf3e0 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Wed, 25 Nov 2020 14:22:26 +0200
Subject: [PATCH] [SW-25565] ppboot: validate image size

We cannot allow large foot print of ppboot as it will increase timing
for BTL to copy into SRAM and by that will japrodize PCIe link
establishment due to breaking 20mS time limit of PRSTn.
goya2_arc_gnu does not have ppboot, therefore validation of size
limitation is not applied with this target.

Change-Id: I4f661938deefc3786e5c85ddb96a4a764ec9638d
Signed-off-by: Ran Shalit <rshalit@habana.ai>
---

diff --git a/Makefile b/Makefile
index 92919e3..4deff31 100644
--- a/Makefile
+++ b/Makefile
@@ -59,7 +59,16 @@
 PPBOOT_OBJECTS += $(PPBOOT_APP_S:.S=.o)
 AUTO_GENERATED_FILE := auto_generated.h
 
-all: bin $(BINARIES) buildversion-habanalabs
+all: prepare_bin image_validation
+image_validation: prepare_bin
+prepare_bin: bin $(BINARIES) buildversion-habanalabs
+
+image_validation_common: ppboot
+	@sz=`stat -Lc "%s" $(PPBOOT_TARGET)`; \
+	if [ $$sz -gt $(PPBOOT_MAX_SIZE) ]; then \
+		echo "$(PPBOOT_TARGET) image size is too big (> $(PPBOOT_MAX_SIZE))"; \
+		exit 1; \
+	fi
 
 $(BINDIR)/%.bin: %
 	$(OBJCOPY) -O binary -R .note.gnu.build-id $(BINDIR)/$< $@
diff --git a/gaudi/Makefile b/gaudi/Makefile
index 431525e..e19dda2 100644
--- a/gaudi/Makefile
+++ b/gaudi/Makefile
@@ -1,4 +1,7 @@
 FW_SIZE := 0x4000
+PPBOOT_MAX_SIZE := 16384
 
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/security.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/reset.c)
+
+image_validation: image_validation_common
\ No newline at end of file
diff --git a/gaudi2/Makefile b/gaudi2/Makefile
index 861ab3b..56b8583 100644
--- a/gaudi2/Makefile
+++ b/gaudi2/Makefile
@@ -1 +1,4 @@
 FW_SIZE := 0x4400
+PPBOOT_MAX_SIZE := 16384
+
+image_validation: image_validation_common
\ No newline at end of file
diff --git a/goya/Makefile b/goya/Makefile
index eacfb1c..1468b5a 100644
--- a/goya/Makefile
+++ b/goya/Makefile
@@ -1 +1,4 @@
 FW_SIZE := 0x8000
+PPBOOT_MAX_SIZE := 6144
+
+image_validation: image_validation_common
diff --git a/goya2/Makefile b/goya2/Makefile
index d0c80b0..8f3266a 100644
--- a/goya2/Makefile
+++ b/goya2/Makefile
@@ -1,2 +1,5 @@
 FW_SIZE := 0x4400
+PPBOOT_MAX_SIZE := 18432
 PREBOOT_SRCS_C += $(shell ls preboot/thermal_v2.c)
+
+image_validation: image_validation_common
diff --git a/goya2_arc/Makefile b/goya2_arc/Makefile
index 9c22872..4b0c6bb 100644
--- a/goya2_arc/Makefile
+++ b/goya2_arc/Makefile
@@ -1,3 +1,6 @@
 FW_SIZE := 0x8000
+PPBOOT_MAX_SIZE := 16384
 
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/pll.c)
+
+image_validation: image_validation_common
